cpp/src/arrow/record_batch.cc

```diff
@@ -248,19 +250,40 @@ Result<std::shared_ptr<StructArray>> RecordBatch::ToStructArray() const {
                                        /*offset=*/0);
 }
 
```

Comment:
(If you would prefer to do this in a follow up that's fine) We have `Float16` now and we can do better than converting HALF_FLOAT to DOUBLE than going through uint16_t:
```suggestion

template <typename Num>
using CTypeOrFloat16 = std::conditional_t<Num::type_id == Type::HALF_FLOAT,
    Float16, typename Num::c_type>;

```

Then below this can be used as
```c++
  template <typename T>
  Status Visit(const T&) {
    if constexpr (is_numeric(T::type_id)) {
      using In = CTypeOrFloat16<T>;
```

... and
```c++
template <typename DataType>
inline void ConvertColumnsToTensor(const RecordBatch& batch, uint8_t* out) {
  using CType = CTypeOrFloat16<DataType>;
```

Reply:
To clarify, we have `Float16` in Arrow but we would need an equivalent c_type, right? 
You mention it would be better to go through `double` then `uint16_t` so I presume I am correct in thinking we do not have an equivalent of `Float16` available. Could you then explain why would it be better to go through `double` than `uint16_t`?

Reply:
Something else about the float16 support. The trick of viewing that data as uint16 is fine for the non-mixed case where we are doing a simple memcpy. But is that still OK when actually casting, as done in this PR? I wouldn't expect so, and in general I don't think we support much casting for float16 (eg at the moment we don't even support casting half_float to float in the cast kernels). 
So maybe we should just raise a NotImplementedError instead when encountering float16 and casting needs to be done?

Reply:
Oh, you are right Joris! I do agree with raising a `NotImplementedError`.

Reply:
Updated in https://github.com/apache/arrow/pull/40359/commits/e4f15a3774980bb3e7dbd9d22064699c8975ca73